File: leetcode/LinkedListTests/LinkedList_82.cs
namespace LinkedListTests;

[TestFixture]
internal class LinkedList_82
{
    class Solution {
        public ListNode DeleteDuplicates(ListNode head) {
            if(head is null) return head;
            var dummyHead = new ListNode(-101);
            dummyHead.next = head;
            var lastUnique = dummyHead;
            while(head is not null) {
                //this is duplicate
                if(head.next is not null && head.val == head.next.val) {
                    while(head.next is not null && head.val == head.next.val)
                    {
                        head = head.next;
                    }
                    lastUnique.next = head.next;
                } else {
                    lastUnique = lastUnique.next;
                }
                head = head.next;
            }
            return dummyHead.next;
        }
    }
}

File: leetcodePython/LinkedList/linkedlist_82.py
from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def deleteDuplicates(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head:
            return head
        dummyHead = ListNode(-101)
        dummyHead.next = head
        lastUnique = dummyHead
        while head:
            if head.next and head.val == head.next.val:
                while head.next and head.val == head.next.val:
                    head = head.next
                lastUnique.next = head.next
            else:
                lastUnique = lastUnique.next
            head = head.next
        return dummyHead.next
